File: sdk/typescript/src/client.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { LeluClient, AuthEngineError } from "../src/index.js";

// ── Mock fetch globally ────────────────────────────────────────────────────────

const mockFetch = vi.fn();
vi.stubGlobal("fetch", mockFetch);

function mockOK(body: unknown) {
  mockFetch.mockResolvedValueOnce({
    ok: true,
    status: 200,
    json: async () => body,
  });
}

function mockError(status: number, errorMsg: string) {
  mockFetch.mockResolvedValueOnce({
    ok: false,
    status,
    json: async () => ({ error: errorMsg }),
  });
}

// Mirrors the engine's POST /v1/agent/authorize response (agentAuthorizeResponse
// in engine/internal/server/server.go). The engine emits boolean flags — there is
// no top-level `decision` field; the SDK derives it from allowed/requires_human_review/compute.
function authorizeResponse(
  decision: "allow" | "deny" | "human_review" | "compute",
  reqId = "req-1"
) {
  return {
    allowed: decision === "allow",
    requires_human_review: decision === "human_review",
    compute: decision === "compute",
    reason: decision === "allow" ? "action allowed" : "action denied",
    trace_id: reqId,
    confidence_used: 0,
  };
}

// ─────────────────────────────────────────────────────────────────────────────

describe("LeluClient", () => {
  let client: LeluClient;

  beforeEach(() => {
    vi.clearAllMocks();
    client = new LeluClient({ baseUrl: "http://localhost:8080" });
  });

  // ── authorize ──────────────────────────────────────────────────────────────

  describe("authorize()", () => {
    it("returns allowed decision", async () => {
      mockOK(authorizeResponse("allow", "t1"));
      const dec = await client.authorize({ tool: "approve_refunds" });
      expect(dec.allowed).toBe(true);
      expect(dec.requestId).toBe("t1");
    });

    it("returns denied decision", async () => {
      mockOK(authorizeResponse("deny", "t2"));
      const dec = await client.authorize({ tool: "delete_invoices" });
      expect(dec.allowed).toBe(false);
    });

    it("returns human_review decision", async () => {
      mockOK(authorizeResponse("human_review", "t3"));
      const dec = await client.authorize({ tool: "wire_transfer" });
      expect(dec.allowed).toBe(false);
      expect(dec.decision).toBe("human_review");
    });

    it("returns compute decision with safeTool and safeArgs", async () => {
      mockOK({
        ...authorizeResponse("compute", "t-compute"),
        safe_tool: "fs.write_file",
        safe_args: { path: "/dev/sandbox/config.yaml" },
      });
      const dec = await client.authorize({
        tool: "fs.write_file",
        args: { path: "/prod/config.yaml" },
      });
      expect(dec.decision).toBe("compute");
      expect(dec.computed).toBe(true);
      expect(dec.allowed).toBe(false);
      expect(dec.safeTool).toBe("fs.write_file");
      expect(dec.safeArgs).toEqual({ path: "/dev/sandbox/config.yaml" });
    });

    it("throws AuthEngineError on HTTP error", async () => {
      mockError(500, "internal error");
      await expect(
        client.authorize({ tool: "approve_refunds" })
      ).rejects.toBeInstanceOf(AuthEngineError);
    });

    it("throws Zod validation error on missing tool", async () => {
      await expect(
        // @ts-expect-error – intentional bad input
        client.authorize({ action: "approve_refunds" })
      ).rejects.toThrow();
    });
  });

  // ── agentAuthorize ────────────────────────────────────────────────────────

  describe("agentAuthorize()", () => {
    it("returns allowed with full confidence", async () => {
      mockOK(authorizeResponse("allow", "t4"));
      const dec = await client.agentAuthorize({
        actor: "invoice_bot",
        action: "approve_refunds",
        context: { confidence: 0.95, actingFor: "user_123" },
      });
      expect(dec.allowed).toBe(true);
      expect(dec.requiresHumanReview).toBe(false);
      expect(dec.confidenceUsed).toBe(0.95);
    });

    it("forwards the actor to the engine (agent_scopes selector)", async () => {
      mockOK(authorizeResponse("allow", "t-actor"));
      await client.agentAuthorize({
        actor: "invoice_bot",
        action: "approve_refunds",
        context: { confidence: 0.95 },
      });
      const body = JSON.parse((mockFetch.mock.calls[0][1] as { body: string }).body);
      expect(body.actor).toBe("invoice_bot");
      expect(body.action).toBe("approve_refunds");
    });

    it("returns requires_human_review at 0.80 confidence", async () => {
      mockOK(authorizeResponse("human_review", "t5"));
      const dec = await client.agentAuthorize({
        actor: "invoice_bot",
        action: "approve_refunds",
        context: { confidence: 0.80 },
      });
      expect(dec.requiresHumanReview).toBe(true);
      expect(dec.allowed).toBe(false);
    });

    it("returns denied decision at low confidence", async () => {
      mockOK(authorizeResponse("deny", "t6"));
      const dec = await client.agentAuthorize({
        actor: "invoice_bot",
        action: "approve_refunds",
        context: { confidence: 0.65 },
      });
      expect(dec.allowed).toBe(false);
      expect(dec.requiresHumanReview).toBe(false);
      expect(dec.traceId).toBe("t6");
    });

    it("validates confidence is between 0 and 1", async () => {
      await expect(
        client.agentAuthorize({
          actor: "bot",
          action: "act",
          context: { confidence: 1.5 },
        })
      ).rejects.toThrow();
    });
  });

  // ── mintToken ─────────────────────────────────────────────────────────────

  describe("mintToken()", () => {
    it("returns minted token", async () => {
      const expiresAt = Math.floor(Date.now() / 1000) + 60;
      mockOK({ token: "jwt.token.here", token_id: "tid1", expires_at: expiresAt });

      const result = await client.mintToken({ scope: "invoice_bot", actingFor: "user_123" });
      expect(result.token).toBe("jwt.token.here");
      expect(result.tokenId).toBe("tid1");
      expect(result.expiresAt).toBeInstanceOf(Date);
    });
  });

  // ── revokeToken ───────────────────────────────────────────────────────────

  describe("revokeToken()", () => {
    it("returns success on revoke", async () => {
      mockOK({ success: true });
      const result = await client.revokeToken("tid1");
      expect(result.success).toBe(true);
    });
  });

  // ── delegateScope ────────────────────────────────────────────────────────

  describe("delegateScope()", () => {
    it("returns delegated token payload", async () => {
      const expiresAt = Math.floor(Date.now() / 1000) + 120;
      mockOK({
        token: "child.jwt.token",
        token_id: "dtid1",
        expires_at: expiresAt,
        delegator: "orchestrator_agent",
        delegatee: "research_agent",
        granted_scopes: ["research"],
        trace_id: "td1",
      });

      const result = await client.delegateScope({
        delegator: "orchestrator_agent",
        delegatee: "research_agent",
        scopedTo: ["research"],
        confidence: 0.92,
      });

      expect(result.token).toBe("child.jwt.token");
      expect(result.tokenId).toBe("dtid1");
      expect(result.grantedScopes).toEqual(["research"]);
      expect(result.traceId).toBe("td1");
      expect(result.expiresAt).toBeInstanceOf(Date);
    });

    it("validates confidence is between 0 and 1", async () => {
      await expect(
        client.delegateScope({
          delegator: "orchestrator_agent",
          delegatee: "research_agent",
          confidence: 1.1,
        })
      ).rejects.toThrow();
    });
  });

  // ── isHealthy ─────────────────────────────────────────────────────────────

  describe("isHealthy()", () => {
    it("returns true on ok status", async () => {
      mockOK({ status: "ok" });
      expect(await client.isHealthy()).toBe(true);
    });

    it("returns false when fetch throws", async () => {
      mockFetch.mockRejectedValueOnce(new Error("connection refused"));
      expect(await client.isHealthy()).toBe(false);
    });
  });
});

File: sdk/go/client_test.go
package lelu

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestAuthorize(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/v1/authorize" {
			t.Fatalf("unexpected route: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"allowed":  true,
			"reason":   "ok",
			"trace_id": "trace-123",
		})
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL})
	res, err := c.Authorize(context.Background(), AuthRequest{UserID: "u1", Action: "view"})
	if err != nil {
		t.Fatalf("authorize failed: %v", err)
	}
	if !res.Allowed || res.TraceID != "trace-123" {
		t.Fatalf("unexpected response: %#v", res)
	}
}

func TestAgentAuthorize(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/v1/agent/authorize" {
			t.Fatalf("unexpected route: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"allowed":               true,
			"reason":                "allowed",
			"trace_id":              "trace-456",
			"requires_human_review": false,
			"confidence_used":       0.95,
		})
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL})
	res, err := c.AgentAuthorize(context.Background(), AgentAuthRequest{
		Actor:      "invoice_bot",
		Action:     "approve_refunds",
		Confidence: 0.95,
	})
	if err != nil {
		t.Fatalf("agent authorize failed: %v", err)
	}
	if !res.Allowed || res.ConfidenceUsed != 0.95 {
		t.Fatalf("unexpected response: %#v", res)
	}
}

func TestMintAndRevokeToken(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/v1/tokens/mint":
			w.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(w).Encode(map[string]any{
				"token":      "jwt-token",
				"token_id":   "tid-1",
				"expires_at": 1700000000,
			})
		case r.Method == http.MethodDelete && r.URL.Path == "/v1/tokens/tid-1":
			w.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(w).Encode(map[string]any{"success": true})
		default:
			t.Fatalf("unexpected route: %s %s", r.Method, r.URL.Path)
		}
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL})
	minted, err := c.MintToken(context.Background(), MintTokenRequest{Scope: "invoice_bot"})
	if err != nil {
		t.Fatalf("mint failed: %v", err)
	}
	if minted.TokenID != "tid-1" {
		t.Fatalf("unexpected token id: %s", minted.TokenID)
	}
	if minted.ExpiresAt.Unix() != 1700000000 {
		t.Fatalf("unexpected expires at: %v", minted.ExpiresAt)
	}

	revoked, err := c.RevokeToken(context.Background(), minted.TokenID)
	if err != nil {
		t.Fatalf("revoke failed: %v", err)
	}
	if !revoked.Success {
		t.Fatalf("expected success=true")
	}
}

func TestIsHealthy(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/healthz" {
			t.Fatalf("unexpected route: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "ok"})
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL, Timeout: 2 * time.Second})
	if !c.IsHealthy(context.Background()) {
		t.Fatalf("expected healthy=true")
	}
}

func TestDelegateScope(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/v1/agent/delegate" {
			t.Fatalf("unexpected route: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"token":          "child.jwt.token",
			"token_id":       "dtid-1",
			"expires_at":     1700000120,
			"delegator":      "orchestrator",
			"delegatee":      "research_bot",
			"granted_scopes": []string{"research"},
			"trace_id":       "td-1",
		})
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL})
	res, err := c.DelegateScope(context.Background(), DelegateScopeRequest{
		Delegator:  "orchestrator",
		Delegatee:  "research_bot",
		ScopedTo:   []string{"research"},
		Confidence: 0.92,
	})
	if err != nil {
		t.Fatalf("delegate failed: %v", err)
	}
	if res.Token != "child.jwt.token" {
		t.Fatalf("unexpected token: %s", res.Token)
	}
	if res.TokenID != "dtid-1" {
		t.Fatalf("unexpected token_id: %s", res.TokenID)
	}
	if res.TraceID != "td-1" {
		t.Fatalf("unexpected trace_id: %s", res.TraceID)
	}
	if len(res.GrantedScopes) != 1 || res.GrantedScopes[0] != "research" {
		t.Fatalf("unexpected granted_scopes: %v", res.GrantedScopes)
	}
	if res.ExpiresAt.Unix() != 1700000120 {
		t.Fatalf("unexpected expires_at: %v", res.ExpiresAt)
	}
}

func TestDelegateScope_Validation(t *testing.T) {
	c := NewClient(ClientConfig{BaseURL: "http://localhost"})
	_, err := c.DelegateScope(context.Background(), DelegateScopeRequest{
		Delegatee: "bot",
	})
	if err == nil || err.Error() != "delegator is required" {
		t.Fatalf("expected delegator required error, got: %v", err)
	}

	_, err = c.DelegateScope(context.Background(), DelegateScopeRequest{
		Delegator:  "orch",
		Delegatee:  "bot",
		Confidence: 1.5,
	})
	if err == nil || err.Error() != "confidence must be between 0 and 1" {
		t.Fatalf("expected confidence validation error, got: %v", err)
	}
}

func TestEngineError(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"error": "unauthorized"})
	}))
	defer ts.Close()

	c := NewClient(ClientConfig{BaseURL: ts.URL})
	_, err := c.Authorize(context.Background(), AuthRequest{UserID: "u1", Action: "view"})
	if err == nil {
		t.Fatalf("expected an error")
	}
	engErr, ok := err.(*EngineError)
	if !ok {
		t.Fatalf("expected *EngineError, got %T", err)
	}
	if engErr.Status != http.StatusUnauthorized {
		t.Fatalf("unexpected status: %d", engErr.Status)
	}
}
